[ResponseOps][Alerting] Flaky test x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/telemetry/index·ts

doakalexi opened the pull request:
Resolves https://github.com/elastic/kibana/issues/140973
Resolves https://github.com/elastic/kibana/issues/136153

## Summary

Fixes a telemetry bug and a flaky telemetry test

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/2717 x 250
https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/2718 x 250

pmuellr commented on x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/telemetry/alerting_and_actions_telemetry.ts:
```diff
@@ -245,49 +213,44 @@ export default function createAlertingAndActionsTelemetryTests({ getService }: F
       expect(telemetry.throttle_time.min).to.equal('0s');
-      expect(telemetry.throttle_time.avg).to.equal('115.5s');
-      expect(telemetry.throttle_time.max).to.equal('600s');
+      expect(telemetry.throttle_time.avg).to.equal('0s');
```
Seems like we should provide some throttle values in the new set of rules, so we can see some non-zero numbers.

doakalexi replied:
Sure, I can add throttle values to one of the rules!

doakalexi replied:
Resolved in this commit [a273697](https://github.com/elastic/kibana/pull/161096/commits/a273697bcdba1a561aea33d302ffdccdb4ff8dfe)

pmuellr commented on x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/telemetry/alerting_and_actions_telemetry.ts:
```diff
@@ -299,12 +262,10 @@ export default function createAlertingAndActionsTelemetryTests({ getService }: F
       // number of rule executions - just checking for non-zero as we can't set an exact number
       // each rule should have had a chance to execute once
-      expect(telemetry.count_rules_executions_per_day >= 21).to.be(true);
+      expect(telemetry.count_rules_executions_per_day >= 0).to.be(true);
```
Unfortunate to have to leave that at potentially zero!  Looking at the code, I think we'd have to call `getEventLog()` on every rule, to make sure it has at least one execute, before requesting the telemetry.  Then we could use the number of rules instead of zero.

doakalexi replied:
This is one of the flaky parts of the test, and I am worried about looking for an exact number of runs. I will remove the >= and just make it >, it should not be 0

pmuellr replied:
heh, no problem!

pmuellr commented on x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/telemetry/alerting_and_actions_telemetry.ts:
```diff
@@ -541,11 +425,11 @@ export default function createAlertingAndActionsTelemetryTests({ getService }: F
+          id: alwaysFiringRuleId[Spaces[2].id],
           provider: 'alerting',
-          actions: new Map([['execute', { gte: 10 }]]),
+          actions: new Map([['execute', { equal: 1 }]]),
```
I think I've wondered about this `equal: n` condition - could it cause more flakiness, if it turns out on the first try it got 0, and on the next try it got 2, so never gets to 1?  Ya, looking a little closer, I think that should be `gte: 1`.

doakalexi replied:
Sure, I can change this

doakalexi replied:
Running in the flaky test runner https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/2740
Through my testing it seemed like this change made the test flaky, I just want to verify

pmuellr reviewed (changes requested):
Generally LGTM.  I'd like to get the `equal` -> `gte` change in the getEventLog() call - that feels like it could someday go flaky.  

The others I'm not so sure about.  Mainly because I'm missing some context.  Could you provide some comments in the PR for:

- why the originating issues were flaky
- why they aren't flaky anymore
- it seems like we removed some rules being tested - I'm curious why

doakalexi commented:
> * why the originating issues were flaky

The original test was flaky from this block of code for both actions and alerting telemetry that is below
```
await retry.try(async () => {
        const telemetryTask = await es.get<TaskManagerDoc>({
          id: `task:Actions-actions_telemetry`,
          index: '.kibana_task_manager',
        });
        expect(telemetryTask!._source!.task?.status).to.be('idle');
        const taskState = telemetryTask!._source!.task?.state;
        expect(taskState).not.to.be(undefined);
        actionsTelemetry = JSON.parse(taskState!);
        expect(actionsTelemetry.runs).to.equal(2);
        expect(actionsTelemetry.count_total).to.equal(20);
      });
```
The number of `actionsTelemetry.runs` was not always 2, and I believe looking for an exact number of runs was causing the  expect(telemetryTask!._source!.task?.status).to.be('idle') to also be flaky where sometimes it would be 'running' instead of 'idle'. This is why I removed looking for an exact count of 2 runs and only wanted to verify that there had been at least one run.
> * why they aren't flaky anymore

I simplified the test, added the object remover, and updated the code block above to remove the flaky parts I mentioned.
> * it seems like we removed some rules being tested - I'm curious why

I removed some rules to help clean up the test, I thought it was a hard to understand and some of the rules seemed to be verifying redundant things. I only kept the minimal amount of rules I needed to verify the telemetry results.

pmuellr reviewed (approved):
LGTM, thanks for changes (and attempts at changes!)
